### PR TITLE
1452 Prepare for new ConfigurationReader and Adapt Templates process

### PR DESCRIPTION
**Pull request by jan-vcapgemini**
Adresses #1452.

@devonfw/cobigen

### CODE REVIEW

File: cobigen/cobigen-core/src/test/java/com/devonfw/cobigen/unittest/util/ConfigurationUtilTest.java [hunk 1]
```diff
@@ -17,33 +17,38 @@
...
       System.setProperty("user.home", userHome.getAbsolutePath());
       Path defaultCobigenHome = userHome.toPath().resolve(ConfigurationConstants.DEFAULT_HOME_DIR_NAME);
-      Path templatesFolder = defaultCobigenHome.resolve(ConfigurationConstants.TEMPLATES_FOLDER);
+      Path templatesFolder = defaultCobigenHome.resolve(ConfigurationConstants.TEMPLATE_SETS_FOLDER);
       Files.createDirectories(templatesFolder);
       String templatesArtifact = "templates-devon4j-1.0.jar";
 
-      Path templatesProject = templatesFolder.resolve(ConfigurationConstants.COBIGEN_TEMPLATES);
-      Files.createDirectories(templatesProject);
+      Path adaptedTemplatesDirectory = templatesFolder.resolve(ConfigurationConstants.DOWNLOADED_FOLDER);
```
Comment (GuentherJulian) [inline cobigen/cobigen-core/src/test/java/com/devonfw/cobigen/unittest/util/ConfigurationUtilTest.java:45]: You refer to the downloaded folder. Wouldn't `downloadedTemplatesDirectory` be a better naming

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java [hunk 2]
```diff
@@ -128,6 +141,31 @@ private static void processJar(Path destinationPath) throws IOException {
 
   }
 
+  /**
+   * Unpacks the source CobiGen_Templates Jar and creates a new CobiGen_Templates folder structure at
+   * $destinationPath/CobiGen_Templates location
```
Comment (GuentherJulian) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java:146]: The location does not match with the new structure of the templates directory

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ConfigurationFinder.java [hunk 3]
```diff
@@ -89,49 +115,46 @@ private static Properties readConfigrationFile(Path cobigenConfigFile) {
...
+    // 3 check adapted folder
+    if (Files.exists(templateSetsAdaptedFolderPath)) {
+      return templateSetsFolderPath.toUri();
+    }
+
+    Path templateSetsDownloadedFolderPath = templateSetsFolderPath.resolve(ConfigurationConstants.DOWNLOADED_FOLDER);
+
+    // 4 check downloaded folder
+    if (Files.exists(templateSetsDownloadedFolderPath)) {
+      return templateSetsFolderPath.toUri();
```
Comment (GuentherJulian) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ConfigurationFinder.java:147]: Couldn't check 3 and 4 be placed in one single if clause?

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java [hunk 4]
```diff
@@ -45,7 +46,17 @@ public static void extractTemplates(Path extractTo, boolean forceOverride) throw
     // find templates will also download jars if needed as a side effect and will return the path to the
     // files.
     URI findTemplatesLocation = ConfigurationFinder.findTemplatesLocation();
-    if (Files.isDirectory(Paths.get(findTemplatesLocation))) {
+
+    Path oldCobiGenTemplatesLocation = Paths.get(findTemplatesLocation);
+
+    if (Files.exists(oldCobiGenTemplatesLocation)) {
```
Comment (GuentherJulian) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java:52]: `findTemplatesLocation` can also return the template-sets folder, right? So don't you need to check in this condition if the returned location is the belongs to the old structure?

File: cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/util/CobiGenPaths.java [hunk 5]
```diff
@@ -75,16 +86,44 @@ public static Path getTemplatesFolderPath(Path home) {
...
+   * @param home cobigen configuration home directory
+   * @return {@link Path} of the template sets home directory
+   */
+  public static Path getTemplateSetsFolderPath(Path home) {
+
+    Path templatesPath = home.resolve(ConfigurationConstants.TEMPLATE_SETS_FOLDER);
+
+    createFolder(templatesPath);
+
     return templatesPath;
```
Comment (GuentherJulian) [inline cobigen/cobigen-core-api/src/main/java/com/devonfw/cobigen/api/util/CobiGenPaths.java:127]: return `createFolder(templatesPath);`

**Review by GuentherJulian — CHANGES_REQUESTED** [review]

File: cobigen/cobigen-core/src/test/java/com/devonfw/cobigen/unittest/util/ConfigurationUtilTest.java (hunk 1 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/test/java/com/devonfw/cobigen/unittest/util/ConfigurationUtilTest.java:45]: Adjusted.

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java (hunk 2 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ExtractTemplatesUtil.java]: Adjusted.

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ConfigurationFinder.java (hunk 3 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/ConfigurationFinder.java:141]: Combined.

**Pull request merged**